### PR TITLE
Improve TypedSet API

### PR DESCRIPTION
**Pull request by jirassimok**
Split from #9841.

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java [hunk 1]
```diff
@@ -135,7 +135,7 @@ public static TypedSet createDistinctTypedSet(
                 false);
     }
 
-    public TypedSet(
+    private TypedSet(
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java:195]: Add rationale to the commit message.

File: core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java [hunk 2]
```diff
@@ -76,6 +76,24 @@ public static TypedSet createEqualityTypedSet(
                 elementHashCodeOperator,
                 elementType.createBlockBuilder(null, expectedSize),
                 expectedSize,
+                functionName);
+    }
+
+    public static TypedSet createEqualityTypedSet(
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java:102]: > Add missing TypedSet factory


what made you think this particular one is "missing"?
add rationale to the commit message.

also, this adds an unused method. it can be added later, when it's actually needed, so i'd drop the commit from here

File: core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java [hunk 3]
```diff
@@ -210,7 +210,7 @@ public boolean contains(Block block, int position)
         }
     }
 
-    public void add(Block block, int position)
+    public boolean add(Block block, int position)
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java:265]: Add
```
/**
 * @return ...
 */

File: core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java [hunk 4]
```diff
@@ -223,7 +223,9 @@ public void add(Block block, int position)
         int hashPosition = getHashPositionOfElement(block, position);
         if (blockPositionByHash.getInt(hashPosition) == EMPTY_SLOT) {
             addNewElement(hashPosition, block, position);
+            return true;
         }
+        return false;
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java:280]: do you know why we don't exist early after `containsNullElement = true;` above?

add unit test coverage for adding null element for the first and second time

File: core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java [hunk 5]
```diff
@@ -193,7 +193,10 @@ private TypedSet(
 
     public long getRetainedSizeInBytes()
     {
-        return INSTANCE_SIZE + INT_ARRAY_LIST_INSTANCE_SIZE + elementBlock.getRetainedSizeInBytes() + blockPositionByHash.size() * Integer.BYTES;
+        return INSTANCE_SIZE
+                + INT_ARRAY_LIST_INSTANCE_SIZE
+                + elementBlock.getRetainedSizeInBytes()
+                + blockPositionByHash.size() * (long) Integer.BYTES;
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java:239]: > Add explicit int/long cast in TypedSet

in the cmt msg, mention why you're doing this

File: core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java [hunk 6]
```diff
@@ -33,6 +33,12 @@
 import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
 
+/**
+ * A set of unique SQL values stored in a {@link Block}.
+ *
+ * <p>Depending on the factory method used, the values' equality may be
+ * determined using SQL equality or {@code IS DISTINCT FROM} semantics.
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java:40]: IS NOT DISTINCT FROM ?

File: core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java [hunk 7]
```diff
@@ -53,8 +59,10 @@
 
     private final int initialElementBlockOffset;
     private final long initialElementBlockSizeInBytes;
-    // size is the number of elements added to the TypedSet (including null).
-    // It is equal to elementBlock.size() - initialElementBlockOffset
+    /**
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java:62]: keep as ordinary comment, since it's a private member

(i know, it's subjective, but the comment already exists, so let's just keep it)

File: core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java [hunk 8]
```diff
@@ -117,6 +145,10 @@ public static TypedSet createUnboundedEqualityTypedSet(
                 true);
     }
 
+    /**
+     * Create a {@code TypedSet} that compares its elements using the semantics
+     * of {@code IS DISTINCT}.
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java:150]: IS NOT DISTINCT FROM ?

File: core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java [hunk 9]
```diff
@@ -133,6 +165,14 @@ public static TypedSet createDistinctTypedSet(
                 functionName);
     }
 
+    /**
+     * Create a {@code TypedSet} that compares its elements using the semantics
+     * of {@code IS DISTINCT}.
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java:170]: as above

File: core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java [hunk 10]
```diff
@@ -212,6 +256,12 @@ public boolean contains(Block block, int position)
         }
     }
 
+    /**
+     * Add the value at the given {@code position} in the given {@code block}
+     * to this set.
+     * @return {@code true} if the value was added, or {@code false} if it was
+     *         already in this set.
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java:263]: my IDE removes indentation here, upon class refrormat

File: core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java [hunk 11]
```diff
@@ -212,6 +256,12 @@ public boolean contains(Block block, int position)
         }
     }
 
+    /**
+     * Add the value at the given {@code position} in the given {@code block}
+     * to this set.
+     * @return {@code true} if the value was added, or {@code false} if it was
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java:262]: add empty line above

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayUnionFunction.java [hunk 12]
```diff
@@ -61,22 +61,24 @@ public static Block union(
...
         int rightArrayCount = rightArray.getPositionCount();
-        TypedSet typedSet = createEqualityTypedSet(type, elementEqual, elementHashCode, leftArrayCount + rightArrayCount, "array_union");
         BlockBuilder distinctElementBlockBuilder = type.createBlockBuilder(null, leftArrayCount + rightArrayCount);
-        appendTypedArray(leftArray, type, typedSet, distinctElementBlockBuilder);
-        appendTypedArray(rightArray, type, typedSet, distinctElementBlockBuilder);
+        TypedSet typedSet = createEqualityTypedSet(
+                type,
+                elementEqual,
+                elementHashCode,
+                distinctElementBlockBuilder,
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/scalar/ArrayUnionFunction.java:69]: Can this impact the order of elements returned from the function? Do we want to tie TypeSet internals to maintain ordering?
Move this commit to a separate PR please, so it can be taken a closer look at.

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayIntersectFunction.java [hunk 13]
```diff
@@ -78,23 +78,23 @@ public Block intersect(
...
 
-        TypedSet rightTypedSet = createDistinctTypedSet(type, elementIsDistinctFrom, elementHashCode, rightPositionCount, "array_intersect");
+        TypedSet rightValues = createDistinctTypedSet(type, elementIsDistinctFrom, elementHashCode, rightPositionCount, "array_intersect");
         for (int i = 0; i < rightPositionCount; i++) {
-            rightTypedSet.add(rightArray, i);
+            rightValues.add(rightArray, i);
         }
 
-        BlockBuilder blockBuilder = pageBuilder.getBlockBuilder(0);
+        BlockBuilder output = pageBuilder.getBlockBuilder(0);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/scalar/ArrayIntersectFunction.java:86]: the original naming is a common pattern. i don't want to bike shed on this, but it feels we're changing something that's subjective into a different, subjective form.

is this at least unifying Array functions implementations between each other? let's add rationale to the commit message, or drop the commit

File: core/trino-main/src/main/java/io/trino/operator/scalar/MapToMapCast.java [hunk 14]
```diff
@@ -235,15 +235,17 @@ public static Block checkBlockIsNotNull(Block value)
     public static Block mapCast(
             MethodHandle keyProcessFunction,
             MethodHandle valueProcessFunction,
-            Type toMapType,
+            Type toType,
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/scalar/MapToMapCast.java:238]: targetType, destinationType ?

File: core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java (hunk 2 above)
Comment (jirassimok) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java:102]: I do use it a few commits later; I can move it there.

File: core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java (hunk 4 above)
Comment (jirassimok) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java:280]: We don't exit early because we still have to add the null if it wasn't already there. We could include a check if null is already contained and exit early, though.

([`TestTypedSet.testGetElementPositionWithNull`](https://github.com/trinodb/trino/blob/fe608f2723842037ff620d612a706900e79c52c8/core/trino-main/src/test/java/io/trino/operator/aggregation/TestTypedSet.java#L83) tests adding multiple nulls.)

File: core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java (hunk 6 above)
Comment (jirassimok) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/TypedSet.java:40]: I phrase it this way because I think it's a bit more consistent than sometimes saying `IS NOT DISTINCT FROM` and sometimes saying `IS DISTINCT FROM`, and I don't think it makes it any less clear, but I could change it.

(And at the nit-picking level, I think phrasing it as "`IS DISTINCT FROM` semantics" covers it; we are using the operator's semantics, though we are reversing its result.)

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayUnionFunction.java (hunk 12 above)
Comment (jirassimok) [reply core/trino-main/src/main/java/io/trino/operator/scalar/ArrayUnionFunction.java:69]: `TypedSet` does maintain insertion order, keeping the first copy of each element added (because it never modifies its block builder besides appending new elements).

I'll split the PR shortly.

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayIntersectFunction.java (hunk 13 above)
Comment (jirassimok) [reply core/trino-main/src/main/java/io/trino/operator/scalar/ArrayIntersectFunction.java:86]: Out of all of these commits, this was the one I was the least sure about, and I almost didn't include it at all; I'll just drop it.

File: core/trino-main/src/main/java/io/trino/operator/scalar/MapToMapCast.java (hunk 14 above)
Comment (jirassimok) [reply core/trino-main/src/main/java/io/trino/operator/scalar/MapToMapCast.java:238]: `targetType` works, or I could also drop this commit.

This one was because when I was reading the function, I kept getting confused and thinking "what's being converted to a map type?"

**Comment by jirassimok** [comment]
Oops, I

**Comment by jirassimok** [comment]
Well, I guess I can't rename the branch.

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayUnionFunction.java (hunk 12 above)
Comment (jirassimok) [reply core/trino-main/src/main/java/io/trino/operator/scalar/ArrayUnionFunction.java]: Split as #9905.

**Comment by findepi** [comment]
@jirassimok build is red

**Comment by findepi** [comment]
@jirassimok let me know when ready for next round

**Comment by jirassimok** [comment]
@findepi I think I've covered every comment.

The last commit is new, and replaces `if (!set.contains(x)) { set.add(x); ... }` with `if (set.add(x)) { ... }`. This seems to be necessary when generating bytecode in `MapTransformKeysFunction`, but the rest of the changes in the commit could be dropped.

**Review by findepi — APPROVED** [review]

**Review by skrzypo987 — APPROVED** [review]
